File: Boj_7569.java
//Main idea: 3차원 bfs
import java.io.*;
import java.util.*;

public class Main {
	private static BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
	private static BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out));
	private static StringTokenizer st;
	private static StringBuilder sb = new StringBuilder();

	
	private static int[][][] boxes;
	private static int N,M,H;
	private static int[] dy = {0,0,1,-1,0,0}, dx = {-1,1,0,0,0,0}, dh = {0,0,0,0,-1,1};
	private static int total;
	public static void main(String[] args) throws IOException {
		tomato();
		out.flush();
		out.close();
	}

	private static void tomato() throws IOException {
		Queue<Point> que;
		que = input();
		int answer = bfs(que);
		if(total != 0)
			answer = -1;
		out.write(String.valueOf(answer));
	}

	private static Queue<Point> input() throws IOException {
		st = new StringTokenizer(in.readLine()," ");
		Queue<Point> tmp = new LinkedList<Point>();
		M = Integer.parseInt(st.nextToken());
		N = Integer.parseInt(st.nextToken());
		H = Integer.parseInt(st.nextToken());
		boxes = new int[H][N][M];
		total = 0;
		
		for(int h = 0; h <H; h++) {
			for(int y = 0; y < N; y++) {
				st = new StringTokenizer(in.readLine()," ");
				for(int x = 0; x < M; x++) {
					boxes[h][y][x] = Integer.parseInt(st.nextToken());
					if(boxes[h][y][x] == 1) tmp.add(new Point(h,y,x));
					else if(boxes[h][y][x] == 0) total++;
				}
			}
		}
		return tmp;
	}
	
	private static int bfs(Queue<Point> que) {
		if(total == 0)
			return 0;
		
		int day = 0;
		total += que.size();
		while(!que.isEmpty()) {
			int size = que.size();
			total -= size;
			while(--size >= 0) {
				Point now = que.poll();
				int h = now.h;
				int y = now.y;
				int x = now.x;
				
				for(int i = 0; i < 6; i++) {
					int nh = h + dh[i];
					int nx = x + dx[i];
					int ny = y + dy[i];
					if(!isInBoundary(nh, ny, nx)) continue;
					if(boxes[nh][ny][nx] == 0) {
						boxes[nh][ny][nx] = 1;
						que.add(new Point(nh,ny,nx));
					}
				}
			}
			day++;
		}
		return day-1;
	}
	
	private static boolean isInBoundary(int h, int y, int x) {
		return(h <0 || y < 0 || x<0 || h >= H || y >=N || x >=M)? false : true;
	}
}

class Point {
	int y,x,h;

	public Point(int h, int y, int x) {
		super();
		this.y = y;
		this.x = x;
		this.h = h;
	}
}

File: Boj_7569.kt
//Main idea: 3차원 bfs
import java.util.*

fun main() {
    val arr = Array(100){Array(100){Array(100){0} } }
    val hArr = arrayOf(1,-1,0,0,0,0)
    val yArr = arrayOf(0,0,1,-1,0,0)
    val xArr = arrayOf(0,0,0,0,1,-1)
    val que = LinkedList<Tomato>()

    var st = StringTokenizer(readLine()!!," ").toList()

    val m = st[0].toString().toInt()
    val n = st[1].toString().toInt()
    val h = st[2].toString().toInt()
    var cnt = 0
    var day = 0

    for(floor in 0 until h){
        for(y in 0 until n){
            st = StringTokenizer(readLine()!!," ").toList()
            for(x in 0 until m) {
                arr[floor][y][x] = st[x].toString().toInt() - 1
                if(arr[floor][y][x] == 0)
                    que.add(Tomato(floor,y,x))
                if(arr[floor][y][x] == -1)
                    cnt++
            }
        }
    }
    while (que.isNotEmpty() && cnt != 0) {
        val tom = que.first()
        que.removeAt(0)
        val f = tom.floor
        val y = tom.y
        val x = tom.x
        for (i in 0..5) {
            val nf = f + hArr[i]
            val ny = y + yArr[i]
            val nx = x + xArr[i]
            if (nf < 0 || ny < 0 || nx < 0 || nf >= h || ny >= n || nx >= m) continue
            if (arr[nf][ny][nx] == -1) {
                arr[nf][ny][nx] = arr[f][y][x] + 1
                day = arr[nf][ny][nx]
                que.add(Tomato(nf, ny, nx))
                cnt--
            }
        }
    }
    if(cnt != 0)
        println(-1)
    else
        println(day)
}
class Tomato(var floor :Int, var y:Int, var x:Int){
}
